File: swift-bridge/Sources/SceneKitBridge/Geometry.swift
import Foundation
import ModelIO
import SceneKit

@_cdecl("scn_geometry_new_box")
public func scn_geometry_new_box(
    _ width: Double,
    _ height: Double,
    _ length: Double,
    _ chamferRadius: Double
) -> UnsafeMutableRawPointer? {
    scnRetain(SCNBox(width: width, height: height, length: length, chamferRadius: chamferRadius))
}

@_cdecl("scn_geometry_new_sphere")
public func scn_geometry_new_sphere(_ radius: Double) -> UnsafeMutableRawPointer? {
    scnRetain(SCNSphere(radius: radius))
}

@_cdecl("scn_geometry_new_cylinder")
public func scn_geometry_new_cylinder(_ radius: Double, _ height: Double) -> UnsafeMutableRawPointer? {
    scnRetain(SCNCylinder(radius: radius, height: height))
}

@_cdecl("scn_geometry_new_cone")
public func scn_geometry_new_cone(
    _ topRadius: Double,
    _ bottomRadius: Double,
    _ height: Double
) -> UnsafeMutableRawPointer? {
    scnRetain(SCNCone(topRadius: topRadius, bottomRadius: bottomRadius, height: height))
}

@_cdecl("scn_geometry_new_plane")
public func scn_geometry_new_plane(_ width: Double, _ height: Double) -> UnsafeMutableRawPointer? {
    scnRetain(SCNPlane(width: width, height: height))
}

@_cdecl("scn_geometry_new_floor")
public func scn_geometry_new_floor() -> UnsafeMutableRawPointer? {
    scnRetain(SCNFloor())
}

@_cdecl("scn_geometry_new_text")
public func scn_geometry_new_text(
    _ string: UnsafePointer<CChar>?,
    _ extrusionDepth: Double
) -> UnsafeMutableRawPointer? {
    guard let string else { return nil }
    return scnRetain(SCNText(string: String(cString: string), extrusionDepth: extrusionDepth))
}

@_cdecl("scn_geometry_new_from_mdl_mesh")
public func scn_geometry_new_from_mdl_mesh(_ meshHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let mesh = scnBorrowMDLMesh(meshHandle) else { return nil }
    let selector = NSSelectorFromString("geometryWithMDLMesh:")
    guard (SCNGeometry.self as AnyObject).responds(to: selector),
          let unmanaged = (SCNGeometry.self as AnyObject).perform(selector, with: mesh),
          let geometry = unmanaged.takeUnretainedValue() as? SCNGeometry
    else {
        return nil
    }
    return scnRetain(geometry)
}

@_cdecl("scn_geometry_first_material")
public func scn_geometry_first_material(_ geometryHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let geometry: SCNGeometry = scnBorrow(geometryHandle) else { return nil }
    if geometry.firstMaterial == nil {
        geometry.firstMaterial = SCNMaterial()
    }
    guard let material = geometry.firstMaterial else { return nil }
    return scnRetain(material)
}

@_cdecl("scn_geometry_set_first_material")
public func scn_geometry_set_first_material(
    _ geometryHandle: UnsafeMutableRawPointer?,
    _ materialHandle: UnsafeMutableRawPointer?
) {
    guard let geometry: SCNGeometry = scnBorrow(geometryHandle) else { return }
    let material: SCNMaterial? = scnBorrow(materialHandle)
    geometry.firstMaterial = material
}

File: src/geometry.rs
use core::ffi::c_void;

use crate::ffi;
use crate::material::Material;
use crate::private::{cstring_from_str, handle_type};

handle_type!(Geometry);

impl Geometry {
    #[must_use]
    pub fn box_geometry(width: f64, height: f64, length: f64, chamfer_radius: f64) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::scn_geometry_new_box(
                width,
                height,
                length,
                chamfer_radius,
            ))
        }
    }

    #[must_use]
    pub fn sphere(radius: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_geometry_new_sphere(radius)) }
    }

    #[must_use]
    pub fn cylinder(radius: f64, height: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_geometry_new_cylinder(radius, height)) }
    }

    #[must_use]
    pub fn cone(top_radius: f64, bottom_radius: f64, height: f64) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::scn_geometry_new_cone(
                top_radius,
                bottom_radius,
                height,
            ))
        }
    }

    #[must_use]
    pub fn plane(width: f64, height: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_geometry_new_plane(width, height)) }
    }

    #[must_use]
    pub fn floor() -> Option<Self> {
        unsafe { Self::from_raw(ffi::scn_geometry_new_floor()) }
    }

    #[must_use]
    pub fn text(text: &str, extrusion_depth: f64) -> Option<Self> {
        let text = cstring_from_str(text)?;
        unsafe { Self::from_raw(ffi::scn_geometry_new_text(text.as_ptr(), extrusion_depth)) }
    }

    /// Wrap a raw `MDLMesh *` pointer and build an `SCNGeometry` from it.
    ///
    /// # Safety
    ///
    /// `mesh` must point to a valid `MDLMesh` object that stays alive for the duration of the call.
    #[must_use]
    pub unsafe fn from_mdl_mesh_raw(mesh: *mut c_void) -> Option<Self> {
        Self::from_raw(ffi::scn_geometry_new_from_mdl_mesh(mesh))
    }

    #[must_use]
    pub fn first_material(&self) -> Option<Material> {
        unsafe { Material::from_raw(ffi::scn_geometry_first_material(self.ptr)) }
    }

    pub fn set_first_material(&self, material: Option<&Material>) {
        unsafe {
            ffi::scn_geometry_set_first_material(
                self.ptr,
                material.map_or(core::ptr::null_mut(), Material::as_ptr),
            );
        };
    }
}
